src/rules.jl

```diff
@@ -88,8 +87,8 @@ _abs_deriv(x) = signbit(x) ? -one(x) : one(x)
 @define_diffrule Base.log(b, x)    = :( log($x) * inv(-log($b)^2 * $b)                          ), :( inv($x) / log($b)                                                       )
 @define_diffrule Base.ldexp(x, y)  = :( exp2($y)                                                ), :NaN
 
-@define_diffrule Base.mod(x, y)    = :( first(promote(ifelse(isinteger($x / $y), NaN, 1), NaN)) ), :(  z = $x / $y; first(promote(ifelse(isinteger(z), NaN, -floor(z)), NaN)) )
-@define_diffrule Base.rem(x, y)    = :( first(promote(ifelse(isinteger($x / $y), NaN, 1), NaN)) ), :(  z = $x / $y; first(promote(ifelse(isinteger(z), NaN, -trunc(z)), NaN)) )
+@define_diffrule Base.mod(x, y)    = :( z = $x / $y; ifelse(isinteger(z), oftype(float(z), NaN), one(float(z))) ), :(  z = $x / $y; ifelse(isinteger(z), oftype(float(z), NaN), -floor(float(z))) )
+@define_diffrule Base.rem(x, y)    = :( z = $x / $y; ifelse(isinteger(z), oftype(float(z), NaN), one(float(z))) ), :(  z = $x / $y; ifelse(isinteger(z), oftype(float(z), NaN), -trunc(float(z))) )
```

Comment:
Maybe a bit simpler:
```suggestion
@define_diffrule Base.mod(x, y)    = :( z = float($x / $y); isinteger(z) ? oftype(z, NaN) : one(z) ), :(  z = float($x / $y); isinteger(z) ? oftype(z, NaN) : -floor(z) )
@define_diffrule Base.rem(x, y)    = :( z = float($x / $y); isinteger(z) ? oftype(z, NaN) : one(z) ), :(  z = float($x / $y); isinteger(z) ? oftype(z, NaN) : -trunc(z) )
```

Reply:
Hmm, some other rules use the ternary operator so I assumed it's fine. But maybe safer to not change too many things here.